client/session.go

```diff
@@ -1224,79 +1230,105 @@ func (s *session) FetchBlocksMetadataFromPeers(
 	nsMetadata namespace.Metadata,
 	shard uint32,
 	start, end time.Time,
 	opts result.Options,
+	isV2 bool,
 ) (result.ShardResult, error) {
 	var (
 		result   = newBulkBlocksResult(s.opts, opts)
-		complete = int64(0)
-		doneCh   = make(chan error, 1)
-		onDone   = func(err error) {
-			atomic.StoreInt64(&complete, 1)
-			select {
-			case doneCh <- err:
-			default:
-			}
-		}
-		waitDone = func() error {
-			return <-doneCh
-		}
-		m = s.streamFromPeersMetricsForShard(shard, resultTypeBootstrap)
+		doneCh   = make(chan struct{})
+		progress = s.newPeerMetadataStreamingProgressMetrics(shard, resultTypeBootstrap)
 	)
 
+	// Determine which peers own the specified shard
 	peers, err := s.peersForShard(shard)
 	if err != nil {
 		return nil, err
 	}
 
+	// Emit a gauge indicating whether we're done or not
 	go func() {
-		for atomic.LoadInt64(&complete) == 0 {
-			m.fetchBlocksFromPeers.Update(1)
-			time.Sleep(gaugeReportInterval)
+		for {
+			select {
+			case <-doneCh:
+				progress.fetchBlocksFromPeers.Update(0)
+				return
+			default:
+				progress.fetchBlocksFromPeers.Update(1)
+				time.Sleep(gaugeReportInterval)
+			}
 		}
-		m.fetchBlocksFromPeers.Update(0)
+	}()
+	defer func() {
+		close(doneCh)
 	}()
 
 	// Begin pulling metadata, if one or multiple peers fail no error will
 	// be returned from this routine as long as one peer succeeds completely
-	metadataCh := make(chan blocksMetadata, 4096)
+	metadataCh := make(chan blocksMetadata, blocksMetadataChBufSize)
+	// Spin up a background goroutine which will begin streaming metadata from
+	// all the peers and pushing them into the metadatach
+	var streamBlocksMetadataErr error
 	go func() {
-		err := s.streamBlocksMetadataFromPeers(nsMetadata.ID(), shard, peers,
-			start, end, metadataCh, m)
-
+		err := s.streamBlocksMetadataFromPeers(
+			nsMetadata.ID(), shard, peers, start, end, metadataCh, progress, isV2)
 		close(metadataCh)
 		if err != nil {
-			// Bail early
-			onDone(err)
+			streamBlocksMetadataErr = err
 		}
 	}()
 
-	// Begin consuming metadata and making requests
-	go func() {
-		err := s.streamBlocksFromPeers(nsMetadata, shard, peers,
-			metadataCh, opts, result, m, s.streamAndGroupCollectedBlocksMetadata)
-		onDone(err)
-	}()
+	// Begin consuming metadata and making requests. This will block until all
+	// data has been streamed (or failed to stream). Note that this function does
+	// not return an error and if anything goes wrong here we won't report it to
+	// the caller, but metrics and logs are emitted internally. Also note that the
+	// streamAndGroupCollectedBlocksMetadata function is injected.
+	streamFunc := s.streamAndGroupCollectedBlocksMetadata
+	if isV2 {
+		streamFunc = s.streamAndGroupCollectedBlocksMetadataV2
+	}
+	s.streamBlocksFromPeers(nsMetadata, shard, peers,
+		metadataCh, opts, result, progress, streamFunc)
 
-	if err := waitDone(); err != nil {
+	// Check if an error occurred during the metadata streaming
+	if streamBlocksMetadataErr != nil {
```

Comment:
You're right I think this is wrong. In practice it probably always works fine because its extremely unlikely that streamBlocksFromPeers would complete inbetween the call to close(metadataCh) and settings streamBlocksMetadataErr, but I'll fix it to use an error channel.

Good catch